cmd/mesh-bootstrap/storage.go

```diff
@@ -0,0 +1,137 @@
+// +build !js
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/0xProject/0x-mesh/keys"
+	"github.com/0xProject/0x-mesh/p2p"
+	p2pcrypto "github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/routing"
+	dht "github.com/libp2p/go-libp2p-kad-dht"
+	peerstore "github.com/libp2p/go-libp2p-peerstore"
+
+	leveldbStore "github.com/ipfs/go-ds-leveldb"
+	"github.com/libp2p/go-libp2p-peerstore/pstoreds"
+	sqlds "github.com/opaolini/go-ds-sql"
+	log "github.com/sirupsen/logrus"
+)
+
+func getPrivateKeyPath(config Config) string {
+	return filepath.Join(config.DataDir, "keys", "privkey")
+}
+
+func getDHTDir(config Config) string {
+	return filepath.Join(config.DataDir, "p2p", "dht")
+}
+
+func getPeerstoreDir(config Config) string {
+	return filepath.Join(config.DataDir, "p2p", "peerstore")
+}
+
+func getNewDHT(ctx context.Context, config Config, kadDHT *dht.IpfsDHT) func(h host.Host) (routing.PeerRouting, error) {
+	switch config.DataStoreType {
+	case leveldbDataStore:
+		newDHT := func(h host.Host) (routing.PeerRouting, error) {
+			var err error
+			dhtDir := getDHTDir(config)
+			kadDHT, err = p2p.NewDHT(ctx, dhtDir, h)
+			if err != nil {
+				log.WithField("error", err).Fatal("could not create DHT")
+			}
+			return kadDHT, err
+		}
+
+		return newDHT
+	case postgresDataStore:
+		newDHT := func(h host.Host) (routing.PeerRouting, error) {
+			var err error
+			sqlOpts := &sqlds.Options{
+				Host:     config.DataDBHost,
+				Port:     config.DataDBPort,
+				User:     config.DataDBUser,
+				Password: config.DataDBPassword,
+				Database: config.DataDBDatabaseName,
+				Table:    "dhtkv",
+			}
+			store, err := sqlOpts.CreatePostgres()
+			if err != nil {
+				log.WithField("error", err).Fatal("could not create postgres datastore")
+			}
+
+			kadDHT, err = p2p.NewDHTWithDatastore(ctx, store, h)
+			if err != nil {
+				log.WithField("error", err).Fatal("could not create DHT")
+			}
+
+			return kadDHT, err
+		}
+
+		return newDHT
+
+	default:
+		log.Fatalf("invalid datastore configured: %s. Expected either %s or %s", config.DataStoreType, leveldbDataStore, postgresDataStore)
+		return nil
+
+	}
+}
+
+func getNewPeerstore(ctx context.Context, config Config) (peerstore.Peerstore, error) {
+	switch config.DataStoreType {
+	case leveldbDataStore:
+		// Set up the peerstore to use LevelDB.
+		store, err := leveldbStore.NewDatastore(getPeerstoreDir(config), nil)
+		if err != nil {
+			return nil, err
+		}
+
+		pstore, err := pstoreds.NewPeerstore(ctx, store, pstoreds.DefaultOpts())
+		if err != nil {
+			return nil, err
+		}
+
+		return pstore, nil
+	case postgresDataStore:
+		sqlOpts := &sqlds.Options{
+			Host:     config.DataDBHost,
+			Port:     config.DataDBPort,
+			User:     config.DataDBUser,
+			Password: config.DataDBPassword,
+			Database: config.DataDBDatabaseName,
+			Table:    "peerStore",
+		}
+		store, err := sqlOpts.CreatePostgres()
```

Comment:
It looks like we are using multiple instances of a datastore. For LevelDB we needed to do this in some cases in order to use a separate directory (LevelDB does not support separate databases). Can we share the same instance of `store` across the codebase? This could also simplify a lot of the code since once we have the `store`, a lot of things are the same.

Reply:
The idea behind new datastore was that for every Key-Value store we use a different table to seperate the key space, which could make things easier and faster for querying data.

If for example we would like to grab provider data from the `dhtStore` (which has only a few entries) we wouldn't have to run the `SELECT` query against a table with hundreds of `peerStore` metadata rows present.

Reply:
I understand the reasoning for using a different table, but that doesn't require multiple connections to the same database. You can access many different tables with a single connection. If this is too much work or isn't supported by `github.com/opaolini/go-ds-sql` I'm okay with creating an issue for it and addressing it later.

Reply:
We can just use multiple datastores using the same `*sql.DB`, added an exported `NewQueriesForTable` in `sql-datastore` here: https://github.com/0xProject/sql-datastore/pull/1/commits/5301c2dd1bf697de5655b58a2f35fc8fdafe0780

Which allows us to:
```go
db, err := sql.Open("postgres", constr)
if err != nil {
  return nil, err
}

dhtStore, _ := sqlds.NewDatastore(db, sqlds.NewQueriesForTable("dhtkv"))
peerStore, _ := sqlds.NewDatastore(db, sqlds.NewQueriesForTable("peerkv"))
```

Reply:
Great. @opaolini that looks good to me.